File: 0284_Peeking_Iterator/solution.java
package cn.huang.leetcode;

import java.util.Iterator;

/*
284. Peeking Iterator
Given an Iterator class interface with methods: next() and hasNext(), design and implement a PeekingIterator that
support the peek() operation -- it essentially peek() at the element that will be returned by the next call to next().

Example:

Assume that the iterator is initialized to the beginning of the list: [1,2,3].

Call next() gets you 1, the first element in the list.
Now you call peek() and it returns 2, the next element. Calling next() after that still return 2.
You call next() the final time and it returns 3, the last element.
Calling hasNext() after that should return false.

Follow up: How would you extend your design to be generic and work with all types, not just integer?

 */
public class LC_0284_PeekingIterator {
    class PeekingIterator implements Iterator<Integer> {
        Iterator<Integer> itr;
        //设一个 peeked 的flag 同时 保存 peeked 过得值.  如果已经peeked过了, next()直接返回保存的值即可.
        boolean peeked;
        int peekVal;

        public PeekingIterator(Iterator<Integer> iterator) {
            // initialize any member here.
            this.itr = iterator;
        }

        // Returns the next element in the iteration without advancing the iterator.
        public Integer peek() {
            if (peeked) {
                return peekVal;
            } else {
                peeked = true;
                peekVal = itr.next();
                return peekVal;
            }
        }

        // hasNext() and next() should behave the same as in the Iterator interface.
        // Override them if needed.
        @Override
        public Integer next() {
            if (peeked) {
                peeked = false;
                return peekVal;
            } else return itr.next();
        }

        @Override
        public boolean hasNext() {
            return peeked || itr.hasNext();
        }
    }
}

File: 0284_Peeking_Iterator/solution.cpp
/*
284. Peeking Iterator

Level: Medium

https://leetcode.com/problems/peeking-iterator
*/
#include <iostream>
#include <vector>
#include <string>
#include <stack>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <numeric>
#include <algorithm>

#include "common/ListNode.h"
#include "common/TreeNode.h"

using namespace std;

/*
  Solution: 
*/

/*
 * Below is the interface for Iterator, which is already defined for you.
 * **DO NOT** modify the interface for Iterator.
 */
struct Iterator
{
	struct Data;
	Data *data;
	Iterator(const vector<int> &nums);
	Iterator(const Iterator &iter);

	// Returns the next element in the iteration.
	int next();

	// Returns true if the iteration has more elements.
	bool hasNext() const;
};

class PeekingIterator : public Iterator
{
private:
	//设一个 peeked 的flag 同时 保存 peeked 过得值.  如果已经peeked过了, next()直接返回保存的值即可.
	bool peeked;
	int peekVal;

public:
	PeekingIterator(const vector<int> &nums) : Iterator(nums)
	{
		// Initialize any member here.
		// **DO NOT** save a copy of nums and manipulate it directly.
		// You should only use the Iterator interface methods.
		peeked = false;
	}

	// Returns the next element in the iteration without advancing the iterator.
	int peek()
	{
		if (peeked)
			return peekVal;

		peeked = true;
		peekVal = Iterator::next();

		return peekVal;
	}

	// hasNext() and next() should behave the same as in the Iterator interface.
	// Override them if needed.
	int next()
	{
		if (!peeked)
			peekVal = Iterator::next();
		peeked = false;
		return peekVal;
	}

	bool hasNext() const
	{
		 return Iterator::hasNext() || peeked;
	}
};